File: research/p384-422/check_certificate.mjs
#!/usr/bin/env node
// Independent BigInt checker. It does not import the Python generator or verifier.
import fs from 'node:fs';
import crypto from 'node:crypto';

const N = 39402006196394479212279040100143613805079739270465446667946905279627659399113263569398956308152294913554433653942643n;
const EXPECTED_SHA256 = '5228f6c12fda873ebb2eecdb34a858a193e92c718efedf81b2c9de4fd72d3d70';

function verify(path) {
  const raw = fs.readFileSync(path);
  if (!raw.length || raw.length > 512000 || raw.at(-1) !== 10 || raw.includes(13)) {
    throw Error('size or LF newline');
  }
  if ([...raw].some(byte => byte > 127)) throw Error('non-ASCII byte');
  const lines = raw.toString('ascii').slice(0, -1).split('\n');
  if (lines.length > 1024) throw Error('operation limit');
  const known = new Set(['1']);
  let previous = 1n;
  let squarings = 0;
  for (const [offset, line] of lines.entries()) {
    const index = offset + 1;
    if (!/^[1-9][0-9]{0,119} [1-9][0-9]{0,119} [1-9][0-9]{0,119}$/.test(line)) throw Error(`row syntax ${index}`);
    const [outputText, leftText, rightText] = line.split(' ');
    if (!known.has(leftText) || !known.has(rightText)) throw Error(`unavailable parent ${index}`);
    const output = BigInt(outputText), left = BigInt(leftText), right = BigInt(rightText);
    if (output <= previous || known.has(outputText)) throw Error(`not strictly increasing ${index}`);
    if (left >= output || right >= output || left + right !== output) throw Error(`incorrect sum ${index}`);
    known.add(outputText);
    previous = output;
    if (left === right) squarings++;
  }
  if (previous !== N - 2n) throw Error('wrong P-384 scalar inversion target');
  const multiplications = lines.length - squarings;
  if (lines.length !== 422 || squarings !== 382 || multiplications !== 40) throw Error('wrong operation counts');
  const digest = crypto.createHash('sha256').update(raw).digest('hex');
  if (digest !== EXPECTED_SHA256) throw Error('certificate SHA-256 differs');
  return {valid:true, operations:lines.length, squarings, other_multiplications:multiplications,
    weighted_0_8:'345.6', target_hex:'0x'+previous.toString(16), sha256:digest};
}

try {
  if (process.argv.length !== 3) throw Error('usage: node verify.mjs certificate.txt');
  console.log(JSON.stringify(verify(process.argv[2])));
} catch (error) {
  console.error(JSON.stringify({valid:false, error:String(error.message)}));
  process.exit(1);
}

File: research/p384-422/check_certificate.py
#!/usr/bin/env python3
"""Independent, dependency-free check of a P-384 n-2 addition chain."""

import hashlib
import json
import re
import sys
from pathlib import Path

N = int("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC7634D81F4372DDF581A0DB248B0A77AECEC196ACCC52973", 16)
EXPECTED_SHA256 = "5228f6c12fda873ebb2eecdb34a858a193e92c718efedf81b2c9de4fd72d3d70"
ROW = re.compile(r"[1-9][0-9]{0,119} [1-9][0-9]{0,119} [1-9][0-9]{0,119}")


def verify(path: Path) -> dict:
    raw = path.read_bytes()
    if not raw or len(raw) > 512_000 or not raw.endswith(b"\n") or b"\r" in raw:
        raise ValueError("size or LF newline")
    lines = raw.decode("ascii").split("\n")[:-1]
    if len(lines) > 1024:
        raise ValueError("operation limit")
    known = {1}
    previous = 1
    squarings = 0
    for index, line in enumerate(lines, 1):
        if ROW.fullmatch(line) is None:
            raise ValueError(f"row syntax {index}")
        output, left, right = map(int, line.split(" "))
        if left not in known or right not in known:
            raise ValueError(f"unavailable parent {index}")
        if output <= previous or output in known:
            raise ValueError(f"not strictly increasing {index}")
        if left >= output or right >= output or left + right != output:
            raise ValueError(f"incorrect sum {index}")
        known.add(output)
        previous = output
        squarings += left == right
    if previous != N - 2:
        raise ValueError("wrong P-384 scalar inversion target")
    multiplications = len(lines) - squarings
    if (len(lines), squarings, multiplications) != (422, 382, 40):
        raise ValueError("wrong operation counts")
    digest = hashlib.sha256(raw).hexdigest()
    if digest != EXPECTED_SHA256:
        raise ValueError("certificate SHA-256 differs")
    return {
        "valid": True,
        "operations": len(lines),
        "squarings": squarings,
        "other_multiplications": multiplications,
        "weighted_0_8": "345.6",
        "target_hex": hex(previous),
        "sha256": digest,
    }


if __name__ == "__main__":
    try:
        if len(sys.argv) != 2:
            raise ValueError("usage: python verify.py certificate.txt")
        print(json.dumps(verify(Path(sys.argv[1])), sort_keys=True))
    except (OSError, UnicodeError, ValueError) as exc:
        print(json.dumps({"valid": False, "error": str(exc)}), file=sys.stderr)
        raise SystemExit(1) from None
